src/Mvc/Mvc.Core/src/ApplicationParts/ApplicationPartManager.cs

```diff
@@ -52,18 +52,57 @@ public void PopulateFeature<TFeature>(TFeature feature)
 
         internal void PopulateDefaultParts(string entryAssemblyName)
         {
-            var entryAssembly = Assembly.Load(new AssemblyName(entryAssemblyName));
-            var assembliesProvider = new ApplicationAssembliesProvider();
-            var applicationAssemblies = assembliesProvider.ResolveAssemblies(entryAssembly);
+            var assemblies = GetApplicationPartAssemblies(entryAssemblyName);
+
+            var seenAssemblies = new HashSet<Assembly>();
 
-            foreach (var assembly in applicationAssemblies)
+            foreach (var assembly in assemblies)
             {
+                if (!seenAssemblies.Add(assembly))
+                {
+                    // Assembly discovery may be duplicated (e.g. if a user explicitly specified an ApplicationPartAAttribute
+                    // that was also codegened), but we do not want duplicate ApplicationParts.
+                    // Avoid duplicates. Note that we prefer doing this over Distinct since the latter isn't
+                    // guaranteed to preserve the original ordering which is important here.
+                    continue;
+                }
+
                 var partFactory = ApplicationPartFactory.GetApplicationPartFactory(assembly);
-                foreach (var part in partFactory.GetApplicationParts(assembly))
+                foreach (var applicationPart in partFactory.GetApplicationParts(assembly))
                 {
-                    ApplicationParts.Add(part);
+                    ApplicationParts.Add(applicationPart);
                 }
             }
         }
+
+        private static IEnumerable<Assembly> GetApplicationPartAssemblies(string entryAssemblyName)
+        {
+            var entryAssembly = Assembly.Load(new AssemblyName(entryAssemblyName));
+
+            // Use ApplicationPartAttribute to get the closure of direct or transitive dependencies
+            // that reference MVC.
+            var assembliesFromAttributes = entryAssembly.GetCustomAttributes<ApplicationPartAttribute>()
+                .Select(name => Assembly.Load(name.AssemblyName))
```

Comment:
So this will result in the assemblies being actually loaded? Is that intended to happen at this phase?
If not, consider using `Assembly.ReflectionOnlyLoad` instead.
If, however this is the intention, the method name should probably be somewhat cleaner - to reflect on that.

Reply:
>  Is that intended to happen at this phase?

Yup, we've already decided that these assemblies are of interest to MVC and will be turned in to `ApplicationPart` instances.